Events based behaviors 🚀

4ian opened the pull request:
Ground breaking feature ahead 🔥

GDevelop has always being based on the idea of allowing anyone to create simple, advanced or very ambitious games. Behaviors have been very helpful as they allow to quickly add features to objects.
I'm taking this one step further by allowing GDevelop users to create their **own reusable behaviors** using the classic **events** system.
* For advanced users, it's a way to factor logic of the objects inside behaviors ("encapsulation"), and to keep a simple, easy to read events sheet in the scene.
* For beginners, this will allow them to reuse behaviors created by other users.

For the full help page, read here: http://wiki.compilgames.net/doku.php/gdevelop5/behaviors/events-based-behaviors 👀
*I would need your help to fix grammar/typos and to tell me the part that are not clear.*

This is still a work in progress but I hope to be able to ship this sooner than later. As far as I know, this is something that was never done in any other game making software with visual programming :)

Still to do:

- [x] Refactoring when behavior function renamed.
- [x] Refactoring when behavior renamed.
- [x] Refactoring when extension renamed (add behaviors and their custom functions)
- [x] Refactoring of expressions when an expression events function is renamed (need to parse expressions, identify behaviors that are of the proper type, and if the function is the one renamed, change its name)
- [x] Prevent renaming a behavior function to a lifecycle method name.
- [x] Prevent the *object type* to be changed if objects of multiple times are using it.
- [x] Copy/paste behaviors and extensions
- [x] Check button to open function works for behavior functions
- [x] Update help buttons

Then asap:

- [x] Easy Import/export
- [x] Sharing/crowdsourced library of behaviors.
- [ ] Work to improve composability of behaviors

4ian commented:
cc @blurymind @ddabrahim @Wend1go @Bouh @PascalLadalle @zatsme or others, if you have a bit of time to read the wiki page, would be very helpful to see if it's properly explained and fix typo/grammar :)
It's quite advanced, but the idea is of course to still have the help accessible for anyone to understand this new feature.

Bouh commented:
WOW !
I compiled this feature to use it, and it's easy to handle, amazing !
Now i need read the wiki for understand every details.

- Some of the current behaviours may have arguments.
I don't see that for personalized behaviours. (Maybe i've miss this point, it's late at home )

  Taking again the Destructible example of the wiki I imagine each instance of my "wall" object with a Destructible behavior and I put a different limit value for each instance in the behavior.
And to go further I could change the limit value for a particular instance via the events.

I've spotted a little thing :
- When we rename behavior the Behavior list, the type for "Parameter#1 Behavior" in parameters for custom function are not refresh.

Wend1go commented:
Pure awesomeness! This will bring object orientated programming to GD. I can hardly imagine all the possibilities this will give us. 🎉 🚀 🎉 

The event based behaviours will be useful for about any game object that has some kind of interactivity.
Personally, I'd build my whole games around this feature.

Disclaimer: I haven't yet tried this branch on my pc.

Some thoughts:
- I'm pretty sure this will become the main part where game logic is defined, so it should be possible to hide the panels to the left and right side for having the full width to work in. (or being able to reduce their width at least)
- A medium to large game will probably have over a hundred behaviours listed inside the "add behaviour to object" dialog. We'd need at least a search box or a tree widget here to keep the list tidy.
- The themes should be adjusted to make the widgets more compact. Like the top row in the event based behaviour editor ("configuration","parameters","object group") and their entry fields.

There are also some grammar mistakes in the wiki article and the lifecycle method descriptions. Will take a look at it when I come home.

blurymind commented:
I am particularly excited about the prospect of community shared reusable objects/behaviors 😸

ddabrahim commented:
Love it, looks great. With custom functions It is going to be a huge game changer! A real hit. Thanks a lot! :fireworks:

Wend1go commented:
Fixed some grammar mistakes in the wiki article.

Another idea for this function:
- Let the user choose an image from the ressources to be used as an icon for the event based behaviour. That would make it easier to differentiate between them.

Wend1go commented on newIDE/app/src/EventsFunctionsExtensionEditor/BehaviorMethodSelectorDialog.js:
```diff
@@ -0,0 +1,186 @@
+            description={
+              <Trans>
+                Events that will be run at every frame (roughly 60 times per
+                second), for every object that have the behavior attached,
```
for every object that **has** the behavior attached

Wend1go commented on newIDE/app/src/EventsFunctionsExtensionEditor/BehaviorMethodSelectorDialog.js:
```diff
@@ -0,0 +1,186 @@
+            description={
+              <Trans>
+                Events that will be run at every frame (roughly 60 times per
+                second), for every object that have the behavior attached, after
```
again "for every object that **has** the bahavior ..."

Bouh commented:
I wanted to do a behavior to create a usable text field.
I created a custom behavior on a text and I wanted via the OnCreated behavior to create a rectangle with an image that encapsulates the text.
But via events it is impossible to create an object other than the object that uses the current behavior.

4ian commented:
> When we rename behavior the Behavior list, the type for "Parameter#1 Behavior" in parameters for custom function are not refresh.

Yup, this should be included in the latest commits :)

> I'm pretty sure this will become the main part where game logic is defined, so it should be possible to hide the panels to the left and right side for having the full width to work in. (or being able to reduce their width at least)

At the least the panel on the right can be closed, and the one on the left reduced. I think it should be enough?

> A medium to large game will probably have over a hundred behaviours listed inside the "add behaviour to object" dialog. We'd need at least a search box or a tree widget here to keep the list tidy.

True, I'll start thinking about how to rework this list. I want at some point to include here a search amongst behaviors that are shared somehow by the community :)

> The themes should be adjusted to make the widgets more compact. Like the top row in the event based behaviour editor ("configuration","parameters","object group") and their entry fields.

You can manually resize the top part to reduce its size, but I agree, I'm not totally satisfied with this for now.

> Fixed some grammar mistakes in the wiki article.

Thanks!!

> Let the user choose an image from the ressources to be used as an icon for the event based behaviour. That would make it easier to differentiate between them.

Yeah at some point I'd like to have custom images. The only thing is that I have to see if I can somehow embed it in the behavior, so that sharing is a matter of copy/pasting the content of the behavior (rather than having to share resources too).

> But via events it is impossible to create an object other than the object that uses the current behavior.

Yes, it's because behaviors don't know anything about your game, and so can't access objects in your game (remember they should be reusable, and are generated like traditional behaviors).
Though you might be able to create a custom function in your behavior (call it Init or something like this) which is an action taking the object to create as parameter. Basically, you're asking the user of your behavior to call by themselves some actions at the beginning of the scene or regularly to pass the object as parameter.
Your behavior stays reusable this way - even if a bit more of a pain to use (because you have to remember to call the proper actions in the events sheet).

4ian commented:
Also concerning customization:

> Some of the current behaviours may have arguments.
> I don't see that for personalized behaviours. (Maybe i've miss this point, it's late at home )

For now these custom behaviors don't have properties (like builtin behaviors can have). I'd like to expose that later, but for now you'll have to use variables of the object. It's less clear but I'll add support for properties later :)
I'll also want to make behaviors more composable, by allowing some behaviors to be plugged in to others. This won't be available immediately, but something that I'd like to work on later.

Bouh commented:
Thanks, it works! And in the control panel, the sentence for the event sheet gives me an error about `_PARAM3_`
I've created a custom function in my behavior. I have 3 Parameters :
```
Parameter #0  Object
Parameter #1  Behavior
Parameter #2  obj_to_create
```
if i wrote `_PARAM3_` for the sentence in event sheet nothing is displayed.

4ian commented:
Can you try the latest commit? I think I've fixed this recently :)

Bouh commented:
> Can you try the latest commit? I think I've fixed this recently :) (in your case, `_PARAM0_` and `_PARAM2_` should be asked for).

After the last commit it's better, thanks you.

I continue the feedbacks with :
 - 🐛 In objects properties the field **Object Name** when we change the name and appy it, the name isn't save and refresh in object list.

- I've created sprite object, my behavior is apply to it. throught the function events i can't modify the size of my sprite object. In the action editor the list of objects is empty.
![image](https://user-images.githubusercontent.com/1670670/57469298-9efbd880-7286-11e9-95ad-d5628e6af7f3.png)

4ian commented:
> I've created sprite object, my behavior is apply to it. throught the function events i can't modify the size of my sprite object. In the action editor the list of objects is empty.

According to what I see on your screenshot, the object type of the behavior is "Any object".
I think it's because when the object type of the behavior is changed *after* a method is created, the method is *not updated* (so in your case to "fix" this, delete and remake a method and it should display "Sprite" as the object type).
I could fix this, but changing the object type of a behavior might be "dangerous" as some object could have been attached to this behavior, and this behavior could not be incompatible (in practice, it's unlikely, but that could happen). Right now, that will crash the preview/exported game if you use a behavior on an object (let's say a text) and then update the behavior to use another object type (let's say a sprite) and launch an action specific to this object type.
(not sure if I'm clear ^^')

I've added this check to my todo list:
> - [ ] Maybe prevent the *object type* to be changed once the behavior is created? Method parameters would need update if we keep allowing this (right now you have to delete them and remake them) but also we could have behaviors associated to objects that are now not valid (and crash the game). Might be easier to ask for the object type and then disallow this to change.

4ian commented:
For the first feedback, you can create a separate bug report I think :) I guess it's a matter of forcing the update/refresh to the list :)

4ian commented:
I've added a button to make it more obvious that a behavior have some options that can be edited, including the object type. It's now safe to edit the object type (GDevelop will gray out types that can't be used, in case the behavior is already used on some objects), functions parameters will be updated.

I've also fixed the copy/cut/past-ing of extensions. With this, it should be much easier to reuse extensions across projects ;)
I'll work later on simplifying a lot sharing, but for now this should be usable.

I'll be less available this week-end, but please test and let me know if you have other questions or feedback if you try this!

Bouh commented:
- When we open behavior we have list of Behavior functions, when we open Functions the panel display "Select a behavior to display the functions inside this behavior" this is a big loss of space.
- New button For edit behavior properties is welcome 👍 
- In the lastest commit  we can now copy/paste a extention
Will it be possible to make copies of functions and behaviors methods?
I mean in a behavior there are methods: will we be able to copy/paste/duplicate them.
And same for functions.
Not sure if am clear.

4ian commented:
Sure, I'll see to add copy/cut/past for functions and behavior methods. :)

For the empty panel, you can close it, but it's true it's not possible if a
behavior is not selected. I'll see what I can do do hide it.


On Fri, May 10, 2019, 18:16 Bouh <notifications@github.com> wrote:

>
>    - When we open behavior we have list of Behavior functions, when we
>    open Functions the panel display "Select a behavior to display the
>    functions inside this behavior" this is a big loss of space.
>    - New button For edit behavior properties is welcome 👍
>    - In the lastest commit we can now copy/paste a extention
>    Will it be possible to make copies of functions and behaviors methods?
>    I mean in a behavior there are methods: will we be able to
>    copy/paste/duplicate them.
>    And same for functions.
>    Not sure if am clear.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/4ian/GDevelop/pull/1049#issuecomment-491364345>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAJYRAQA3IR247V25JRVMG3PUWUXXANCNFSM4HLNFHOQ>
> .
>

4ian commented:
I've added copy/cut/paste for functions and behavior methods :)

blurymind commented:
Can we save to a file and load from a file too? This is getting exciting :D

On Sun, May 12, 2019 at 11:38 PM Florian Rival <notifications@github.com>
wrote:

> I've added copy/cut/paste for functions and behavior methods :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/4ian/GDevelop/pull/1049#issuecomment-491635276>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABRRWVLMURHMV3LKQ6WBP63PVCL6FANCNFSM4HLNFHOQ>
> .
>

4ian commented:
> Can we save to a file and load from a file too? This is getting exciting :D

Well, everything that you copy/paste in GDevelop is actually JSON ;) You can certainly paste the content in a file, just hit Ctrl+V (or Cmd+V) in a text editor:
 
![image](https://user-images.githubusercontent.com/1280130/57656860-86b2f300-75d1-11e9-8285-b8d75f7117a2.png)
(here, an extension as an example, mostly empty, that I copied from GD)

then when you when to import it, you copy the content from the file and you'll be able to paste it in GDevelop as usual (technically, GDevelop does a check on the beginning of the string, here `"000kind":"GDEVELOP_Events Functions Extension_CLIPBOARD_KIND-jsBdHbLy912y8Rc"`, to recognize things that can be pasted in GDevelop).

Not as convenient as an import/export to files, but I'm thinking of other ways to enable sharing all of this ;) (if you have ideas, they are welcome)

Bouh commented:
I work to add input field behavior and i've created new conditions with the functions, like 
- if a Global variable exist
- if a scene variable exist
- if a object variable exist 
in the functions we can get the value of parameters with `GetArgumentAsString()` and `GetArgumentAsNumber()`
It's possible to add  `GetArgumentAsObject()` ?
Because actually i use  `GetArgumentAsString()` for pass the name of the object and get in JS with `runtimeScene.getObjects()` In the events it's fine, but not in custom behavior. 
Because all objects with the same name are returned, and i wish just the object.
I want pass the object from my behavior and i can't. 

In my case my i try to do custom behavior, i've a condition check if the object have the object variable "Text_var", if this variable contains a stringthe object display the string.

Phenomena3 commented:
I'm excited to experiment with this! Sounds great :)

blurymind commented:
If it's really that easy, it should really be trivial to be able to load
and save addons from json files. How about we have gdevelop automatically
scan the project/extensions folder for json files when loading the project,
and if they are formatted properly, have them loaded.

The idea is to make it conveniently simple to load extensions from files
that have been shared by the community.
Godot uses that approach and it is very simple, as it requires no steps.

The tricky part is saving changes back to the files. If an extension has
been loaded from a file, GD should keep track of its path and write changes
back to it whenever the project is saved

On Tue, May 14, 2019 at 5:08 AM Phenomena3 <notifications@github.com> wrote:

> I'm excited to experiment with this! Sounds great :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/4ian/GDevelop/pull/1049?email_source=notifications&email_token=ABRRWVJ2N22JGTV7BVK5VGTPVI3M5A5CNFSM4HLNFHO2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVKGRFI#issuecomment-492071061>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABRRWVKWE7F7FY25QYLCMFLPVI3M5ANCNFSM4HLNFHOQ>
> .
>

4ian commented:
> They even have a browser that shows approved extensions shared at the repository and downloads them automatically to that folder when the user likes to use them in his project

I'm building a very simple version of this: https://github.com/4ian/GDevelop-extensions 😉
It's super early but you get the idea: a list of extensions, that will be able to be searched from GDevelop and installed in one click in a project (I say "Extensions", because the advantage of sharing an entire extension is that you can have a function, or a behavior, or even bundle multiple of them). 

I'd like to make it easy to share these extensions containing behaviors, but for now to go as fast as possible I'll ask to submit PRs or just to contact me and give me the extension json file.
In any case, at the beginning I need to curate this list to be sure to provide high quality behaviors (or functions) :)

If you have ideas about useful "events based" behaviors or already started making some, please send them to me here :D (or open an issue/PR on the other repository)

blurymind commented:
This is a fantastic start! Maybe we can have some sort of an IDE command to submit an extension for review- similar to the one for submitting bugs?

ddabrahim commented:
Yeah, since it is only a JSON string it would be nice if we could submit and share right from GD, however review everything people going to submit would be too much for anyone. I think would be the best to allow to share anything and let us vote on extension and sort them by popularity "most downloaded" and "most votes" also have some flagged as "featured" maybe, the really good ones. Don't know how much would it cost on the server side but could make it a subscribers only features to access and download to cover some expenses to let us upload and the maintenance of the system to vote and sort by popularity..etc. Could also let Discord and Forum mods to also moderate the list maybe if they are interested or anyone who is considered a "trusted" member...

Regarding what behaviors to include. I'm sure many of the examples demonstrating a single or couple feature only, could be turned in to an extension like "vehicle controls", "camera shake", "button", "text to speech", "joystick control", "top-down chasing AI", "zombie-AI (of course)" "platformer-AI" ..etc Even templates, complete game mechanics could be turned in to behaviors. The nice thing about it, anything can be done with this, we could essentially turn GD in to a "maker" type of engine for any genre, which I'm looking forward to do.

However, unfortunately  I don't have time at the moment to make some for the initial release, but definitely going to share some after release. For the initial release, could be enough just a few example and let the community take care of extending the list. I did begin to turn some examples in to custom functions a while ago but then did not find the time to finish it, I think these could be turned in to a behaviors easily and could be also useful:
https://github.com/GameTemplates/GDevelop-Examples/tree/master/custom-functions

Bouh commented:
It's easy to share some functions too, I have a few, I put you the functions copied from the clipboard:
Conditions : 
[Global variable exist](https://pastebin.com/raw/TZuBV40p)
[Scene variable exist](https://pastebin.com/raw/4nSRfaNu)

Actions  : 
[FormatTime seconds to HH:MM:SS ](https://pastebin.com/raw/iSs4RaW9)
[Paste content from clipboard in scene variable](https://pastebin.com/raw/UCk8qknp)

It's better i think to share this format if we want just one action/conditions instead download the extention with all fonctions.
I hate download a package for just use 1% of it ^^

4ian commented:
> however review everything people going to submit would be too much for
anyone

Indeed, this can become a lot. Also, we'll probably have tons of submission
that are not that useful/relevant.
So I do intended to have some ranking based on download number/votes and
also some "featured" extensions.

For now as this can be a lot of development and I would like to release
this not too late, I'll go with an official GitHub repository: you can
think of it as a list of featured, easy to download extensions :) You'll
have to submit an issue or a PR to add an extension, but I think this
barrier is not too hard (there will be a button in GDevelop anyway) and
still high enough to be sure that people that want to submit an extension
are taking the time to provide something that is good quality.

> complete game mechanics could be turned in to behaviors

Yes I think so :)

> The nice thing about it, anything can be done with this, we could
essentially turn GD in to a "maker" type of engine for any genre, which I'm
looking forward to do.

This is what I look forward too. The next step will be to ease/allow
composability (behaviors that rely on other behaviors) to allow assembling
behaviors into more complex ones easily.

>
https://github.com/GameTemplates/GDevelop-Examples/tree/master/custom-functions

Nice! I'll take a look :)

> It's easy to share some functions too, I have a few, I put you the
functions copied from the clipboard:

Thanks too!

I'll add both a way to easily find extensions from the "new behavior
dialog" as well as directly from the project manager.
I'll try to bundle your existing functions/behaviors into the repo - please
ping me if I don't do it, it will be easy to add them once I have the whole
thing set up and the new version of GD released :)

> I hate download a package for just use 1% of it ^^

No worries about this, if this become a really large problem in the future
I could generate code only for functions that are actually used ;)

In any cases, glad to hear your interest :) Pretty sure once this is ready,
we'll start having more and more interesting and powerful
functions/behaviors.

On Fri, May 17, 2019 at 4:17 PM Bouh <notifications@github.com> wrote:

> It's easy to share some functions too, I have a few, I put you the
> functions copied from the clipboard:
> Conditions :
> Global variable exist <https://pastebin.com/raw/TZuBV40p>
> Scene variable exist <https://pastebin.com/raw/4nSRfaNu>
>
> Actions :
> FormatTime seconds to HH:MM:SS <https://pastebin.com/raw/iSs4RaW9>
> Paste content from clipboard in scene variable
> <https://pastebin.com/raw/UCk8qknp>
>
> It's better i think to share this format if we want just one
> action/conditions instead download the extention with all fonctions.
> I hate download a package for just use 1% of it ^^
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/4ian/GDevelop/pull/1049?email_source=notifications&email_token=AAJYRAWERVDXAEZYOYKUDBTPV3EB3A5CNFSM4HLNFHO2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVVBNVQ#issuecomment-493491926>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJYRAW4DTHYUC46QGI3TH3PV3EB3ANCNFSM4HLNFHOQ>
> .
>

blurymind commented:
perhaps we could do it similarly to the linux distro packaging model - you have a curated channel with official addons- but you also have a community channel with any casual addons.

Once in a while based on  popularity/demand a community package could become curated - but it must pass review

blurymind commented:
Maybe one thing we can do some time in the future is have something similar to pico8's BBS board
https://www.lexaloffle.com/bbs/?cat=7&carts_tab=1#mode=carts&sub=2

Users can upload any content to it - some cartridges are reusable code, others are graphics or music.A lot of the cartridges are tiny games that can be played straight in the browser.

What is nice about GD is that due to it's html5 nature, we can not only do that - but have it seamless straight inside the the ide. The whole process of sharing one or many resources you drew with piskel inside of GD's newIde to gdevelop's community board. Or function/behaviour you made, even along with an example scene where it's used that can be played in the browser. Just using a "Share" button and your gdevelop account.
Then others can find it in the community board. If they don't like it - report/downvote it. If they like it - import it in a single click to their project and try it.
If they want to submit their improvement - submit it back as a fork of the original.
Along with the community board- another channel - the curated content channel. Everything approved by GD devs that has been tested well. A board where only the best of the best make it :)

Anyways, just dreaming big here :D 

GD becoming an indie dev scene content powerhouse one day

ddabrahim commented:
> as this can be a lot of development and I would like to release this not too late, I'll go with an official GitHub repository

I don't doubt it could require lot of your free time to develop, but basically we are talking about nothing but text. It is not meant to store any valuable resources like user names, passwords, payment info..etc but text and number to count the number of votes for each entry. Could be possible to setup some basic, temporary SQL database somewhere, anywhere. There are many free hosts out there. It could be done in few hours just to get something up and running and hook GD up to it with a submit button and a vote button. Just temporary to have something in place.

> enough to be sure that people that want to submit an extension are taking the time to provide something that is good quality.

Depends, some people may provide a basic framework that may seems useless for many but for some may provide a foundation that they can build on top then they also share what they got and eventually we end up with something that useful for many.

Wend1go commented:
@blurymind I like your idea. Maybe we could add some kind of "Steam Green Light" to it.
So everything that gets uploaded is visible from the website but only behaviours that reached a certain number of :+1: are shown inside the IDE itself. This would keep the quality up and only has a small administrative overhead. New users of GD would get a better impression of GD when only the best stuff is shown.

4ian commented:
You're all having too much ideas compared to the speed I can build things 😄Seriously speaking, I agree with most of what you suggest. In an ideal world, we would have a full featured sharing system with automatic upload, votes, favorites, featured extensions and a powerful search.

For now, here is a quick overview of the sharing capabilities:

Search for new behavior directly from the "New behavior dialog":
![image](https://user-images.githubusercontent.com/1280130/57972373-3cdf5980-7991-11e9-8830-ebf31d1faa70.png)


A similar dialog to search for extensions, accessible from the project manager, or import an extension from file:
![image](https://user-images.githubusercontent.com/1280130/57972343-effb8300-7990-11e9-8d3b-fd4723c4357c.png)

Extensions have now a button to be exported to a file, and also a button to create an issue on GDevelop-extensions repository with a template to fill:
![image](https://user-images.githubusercontent.com/1280130/57972358-128d9c00-7991-11e9-9df8-f4939d76e2e4.png)

Extension descriptions are displayed before installing one. Markdown is supported for description, will should enable nice presentation of real extensions:

![image](https://user-images.githubusercontent.com/1280130/57972389-70ba7f00-7991-11e9-9394-3190c4450045.png)

Extensions can also be searched by "tags". Will be useful when we start to have tons of extensions :)
We can imagine tons of things on top of this basis:
* featured extensions
* make new extension more visible
* sort extensions by downloads, popularity, newest first, etc...
* votes
* automatic upload linked to your GDevelop account

As with everything, more complex systems = longer to build, harder to maintain. I'm starting with the git repository as it's publicly searchable, open-source, has support for discussions when submitting an extension (through the issues), has near 100% availability, automatic backups, browseable with just an internet browser, infinite scalability for a cost of zero, and moderation consists in approving issues/pull requests (rather than having to delete potential extensions containing illegal stuff 🙃Yeah that could happen). It's also a good basis for the future features based on it and it's easy to migrate to something else later like a "real" DB.

We'll also have more challenges to overcome to make the sharing of extensions reach its full potential but I'm pretty sure we're on a very good start.

I should merge this branch soon and start working to add behavior/functions to the "registry".

@Bouh:

> Actions :
> FormatTime seconds to HH:MM:SS 
> Paste content from clipboard in scene variable

Could you slightly rework these so that these are expressions? I would rather avoid using scene variables because there is no official support for them in functions right now.
For the "featured" extensions that we'll release with the next versions, I prefer to stick with something that is not breaking the conventions of the actions/conditions/expressions of GDevelop :)

4ian commented:
Note also that exporting/importing extensions as files allow for an "organic grow" of extensions (read: people exchanging/reusing functions using files). This is very fine and will be also a good metric of the success of the Events based behaviors and extensions :)

blurymind commented:
@4ian it's starting to look fantastic! 😮  One suggestion for the wording of

> Submit to the community extensions

can be

> Submit extension to the community

if

> Edit extension options

is just a header, it could just say

> Extension options

Just suggestions :)

Bouh commented:
I rework my functions to put them into expressions, I have the impression that the action "Set return value to text" doesn't work.
**UTB__ReturnTimeFormated** in debugger is "02:00"
![image](https://user-images.githubusercontent.com/1670670/57983612-5e554980-7a54-11e9-821d-e1ebc19ea7ad.png)

Bouh commented:
I have a behavior and in it I have a condition that I created with the functions.

My condition is "If object X has an object variable named Y"
X is the object of my behavior, type of X is Text.
I can retrieve Y with GetArgumentAsString() and pass it to a temporary variable to manipulate Y in Javascript afterwards.
For X I can't get it back, nor can I manipulate it in Javascript.

4ian commented:
> I rework my functions to put them into expressions, I have the impression that the action "Set return value to text" doesn't work.

There was an error indeed, the parameter shown in the IDE is shown as a (number) expression, while it should be a string (so in your case, you must use VariableString).
I've fixed this (once latest version of libGD.js is compiled, you should get the fix too). You can also ignore the errors that you get: the IDE will complain but code generation will be made properly.

Bouh commented:
- Thank for fix, it's working now :)

- I have a suggestion, i have a behavior and in it I have a condition that I created with the functions.
My condition is "If object X has an object variable named Y"
X is the object of my behavior, type of X is Object (Sprite).
I can retrieve Y with GetArgumentAsString() and pass it to a temporary variable to manipulate Y in Javascript afterwards.
For X I can't get it back, nor can I manipulate it in Javascript.

4ian commented:
@Bouh: you can do this already I think by choosing "Object" as the objects to pass to the JS function:
![image](https://user-images.githubusercontent.com/1280130/57986746-d2054f80-7a70-11e9-840e-bad9d5b67ab3.png)

Complete example:
![image](https://user-images.githubusercontent.com/1280130/57986748-daf62100-7a70-11e9-8a3c-1ce3c86b116d.png)

At least this is working for me:
![image](https://user-images.githubusercontent.com/1280130/57986743-ca45ab00-7a70-11e9-92ff-8e99d7b35bab.png)

blurymind commented:
damn I wish we could import libraries from the project folder easily. Seeing you guys write extensions like that straight into the embedded monaco editor makes it more and more apparent that is the way forward with how we write and distribute extensions. If I could just write 
`import {someClass} from 'myFolder/someLibrary.js'` 
in there.. while `myFolder` is placed in the root of the project folder. Then when you export the extension, it could somehow bundle all the imported libraries within the extension file itself.

4ian commented:
I see the idea, though I don't think that would go through import/require support (that would need GDevelop to actually parse the JavaScript, something that you probably don't want at all, or at least complex bundling with webpack/other bundlers). But I see the idea of having external files bundled with the project.
For now though, let's see how things are going with these events backed extension. The idea is still to make the most of events - which is GDevelop primary "language" :) 
If lots of extensions are implemented using JS, we can think again about it (but also keep in mind that if later GDevelop has a new export based on a native language like C++/Kotlin/Swift/Objective C/Java/WebAssembly/whatever, all these extensions using JS would not work).

4ian commented:
This should be good enough to merge and have this in the next version :) 
I think we'll need quickly to add properties to the events-based behaviors (properties that can be set form the object editor), I've created a card on the roadmap for this: https://trello.com/c/lEpgrCPN/283-allow-custom-behaviors-to-have-properties-that-can-be-set-in-the-object-editor

4ian commented:
I have a few ideas of simple yet useful behaviors, for example:

* Animated counter text (when set on a text, the text will animate itself to go from 0 (or whatever is the current value) to the new value). Can be done using a variable.
* Making an object flash (possible property in the future: speed of the flash)
* Sine movement (would need to have properties, because lots of things can be customized: direction, amplitude of the movement, speed).
* Behavior that force an object to stay on screen (should be doable without properties - but we could actually adding some margins later)
* Objects that move back and forth (property in the future: speed, initial direction)
* Objects with health/damages
(feel free to give more ideas - I think we could even have a custom column for this in the roadmap! What do you think?)

Most of them would greatly benefit from having properties to customize them to be really useful. For now we can start with variables, but I'll try to see soon if I can add support for properties for these behaviors :)

Wend1go commented:
Cool, can't wait to try this. 😍 
I just hope it won't lead into an uncontrolled mess when JS is used in the custom behaviours they might need to be checked/updates when we update any library GD relies on like Pixi or electron. 
Btw, is it possible to add functions to existing extensions? Like adding a function to change the opacity/object colour of Panel Sprites or Tiled Sprites?
I have been wanting to add this for a long time now but always shied away cause it is still in the C++ part of the engine.

4ian commented:
> I just hope it won't lead into an uncontrolled mess when JS is
> used in the custom behaviours they might need to be checked/updates
> when we update any library GD relies on like Pixi or electron.

Yes that's a risk, this is why I warned against using too much JS in my
previous message. As much as possible:
* Behaviors should be done using plain old events - because they are
guaranteed to work across GDevelop versions :)
* Then, if a behavior really needs JS, then it should rely as much as
possible on well documented methods from GDevelop game engine (see
http://4ian.github.io/GD-Documentation/GDJS%20Runtime%20Documentation/)
* Then, if a behavior really needs to use Pixi.js or Electron specific
code, then it can be done but should I guess be stated in the description
("this will only work for Windows/macOS/Linux" for example for Electron)
and should use something like `if (typeof electron !== 'undefined') { ...
}` in the code to avoid crashing the game if not running with Electron.

Note that for now I'll review the submitted extensions so I'll double check
that. In the future, it should also be quite easy to just update this
script (
https://github.com/4ian/GDevelop-extensions/blob/master/scripts/update-extensions-registry.js)
to analyze extensions: if "electron" is found, or there are some usage of
JS, this can be displayed in GDevelop with some badge ("Windows/macOS/Linux
only").
In the future, if extensions can be submitted without my review, we'll
still verify the "featured" extensions to make sure they follow these
guidelines (events as possible, JS if needed, but using proper checks and
public methods only).

Also in case I change something in the game engine that will break
compatibility with some extensions using JS, I'll try to update the related
community extensions - but this will be more of a headache for sure.

> Btw, is it possible to add functions to existing extensions? Like adding
a function to change the opacity/object colour of Panel Sprites or Tiled
Sprites?

So for downloaded extensions yes because they are simply added to your
project (but if you update/re-download the extension, any changes will be
erased).
For existing "builtin" extensions, not directly but you can totally create
a behavior that is limited to an object type (select the object type in the
behavior properties, see
http://wiki.compilgames.net/doku.php/gdevelop5/behaviors/events-based-behaviors).
This allow you to tell that you're behavior is only working with objects
that are for example Panel Sprites or Tiled Sprites. Actually, you could
think of existing objects as being a set of behaviors that have been merged
together :)
If something cannot be done using events, you can then use JS (to access
for example to the Pixi.js rendered object and modify the opacity).

Note that if an extension is using JS and is really useful or simple, I
could help and integrate your code directly in the game engine :) But the
idea is that now with custom behaviors, you can do it without me ;)
That's why using JS inside custom events is interesting (though again, if
you can use only events please do! :))

Look forward to having people create behaviors :D Hopefully I can release a
new version not too late with this.



On Mon, May 20, 2019 at 7:14 AM Wend1go <notifications@github.com> wrote:

> Cool, can't wait to try this. 😍
> I just hope it won't lead into an uncontrolled mess when JS is used in the
> custom behaviours they might need to be checked/updates when we update any
> library GD relies on like Pixi or electron.
> Btw, is it possible to add functions to existing extensions? Like adding a
> function to change the opacity/object colour of Panel Sprites or Tiled
> Sprites?
> I have been wanting to add this for a long time now but always shied away
> cause it is still in the C++ part of the engine.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/4ian/GDevelop/pull/1049?email_source=notifications&email_token=AAJYRAQ2FKMHHFQ7K22IAITPWI6UZA5CNFSM4HLNFHO2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVXZJ5Y#issuecomment-493851895>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJYRAS3LLMPXIXXFN74NZ3PWI6UZANCNFSM4HLNFHOQ>
> .
>

ddabrahim commented:
> I've also added a second function by @ddabrahim...(snap to grid). 

Actually, "snap to grid" was done by @Lizard-13, I only implemented his work in the function so the credit goes to him.

4ian commented:
All, for your interest:
* I've added a Flash behavior (to make an object blink)
* Also pushed some fixes on master to properly allow behavior methods to call other methods from the behavior.

Still a few things that I'd like to iron out before a release.

> Actually, "snap to grid" was done by @Lizard-13, I only implemented his work in the function so the credit goes to him. Actually, I prefer not to take credit for any of that. All done by others.

Right, I changed to make him the author. If you prefer, I can just credit "GDevelop Team" for extensions that don't have a proper owner.

Bouh commented:
Dialog windows have a problem We can't use buttons if the description is too long :|
![image](https://user-images.githubusercontent.com/1670670/58098285-7d311880-7bd9-11e9-9b41-3bfc5f0f0676.png)

- @4ian It works great thanks, this feature is underutilized :o 
- I have a question about variable naming. If names are identical, GD could mix the values? What do you recommend on variable naming, because in the future the behavior/functions list will necessarily have variables that will collide.

I rewrote the conditions in expressions, and I create new ones, I submitted for review on the repo of extensions, [link](https://github.com/4ian/GDevelop-extensions/issues/2).

ddabrahim commented:
> If you prefer, I can just credit "GDevelop Team" for extensions that don't have a proper owner.

I would go with "GDevelop Community" considering if the origin of something is uncertain, probably it has been shared by 'someone' on Discord or the Forum. But I leave it up to you, I don't mind.

Thanks.

4ian commented:
@Bouh:

> @4ian It works great thanks, this feature is underutilized :o

Which one do you mean?

> I have a question about variable naming. If names are identical, GD could mix the values? What do you recommend on variable naming, because in the future the behavior/functions list will necessarily have variables that will collide.

For now I recommend to use object variables as much as possible (or scene variables if needed), but prefix them by something like the name of the extension, or some acronym (for example Flash_Duration for the duration of the flash, and not just Duration).

Ideally in the future, we'll have variables that are specific to Behaviors or even to the current functions - but not soon so for now let's prefix the variables to avoid collisions between variables of different behaviors (or even to avoid confusing the user that could "break" a behavior by misuing a variable).

> We can't use buttons if the description is too long

Can you reproduce it with some game? I can't manage to do it by myself.

Bouh commented:
> Which one do you mean?

Passing an object in javascript events.

>Can you reproduce it with some game? I can't manage to do it by myself.

I paste "Lorem ipusm" text in fields and button gone far because text area have no limit.
https://youtu.be/i7_QvJw1Gyo

blurymind commented:
looks like a css problem :) max-width +overflow:auto? :)

4ian commented:
Material-UI components, here TextField, should already take care of that. I think the fix is adding `rowsMax` , so that the text field start scrolling its content after being too large. This is what was done for the Text Editor, seems to work properly.
Otherwise, the Material-UI Dialog is confused because of the content that is changing its size.

4ian commented:
https://github.com/4ian/GDevelop/releases/tag/v5.0.0-beta69 👀

Let me know if there is any major issue!

Bouh commented:
I can't copy a behavior, I must have misread a commit I thought it was possible :/
